### PR TITLE
Public Markdown-inline API: atomic spans, sentence spans, AST/links

### PR DESCRIPTION
**Pull request by jlevy**
## Summary

Adds a small, general-purpose **public inline API** so any downstream consumer can reuse flowmark's Markdown-inline knowledge instead of copying internals. Two new public submodules (`flowmark.atomic`, `flowmark.ast`); top-level re-exports kept conservative (`Link`, `extract_links`).

Originally opened as a planning spec; now contains the full implementation (Phases A–C from `docs/project/specs/active/plan-2026-05-26-public-inline-api.md`).

### `flowmark.atomic`
- Publishes the atomic-construct patterns + `AtomicPattern` (its delimiter fields now default to `""`, so it's a usable public type for custom constructs).
- New `AUTOLINK` + `BARE_URL` patterns; `MARKDOWN_INLINE_PATTERNS` (code spans, links, autolinks, bare URLs) — purpose-built for prose, deliberately not a subset of the wrapping set.
- `iter_atomic_spans` (offset-preserving, selectable patterns, round-trips) and `iter_atomic_words` (glues atomic constructs, keeps offsets) — now the single tokenization primitive; the wrapping word splitter is reimplemented on it.
- `split_sentences_with_spans` (atomic-aware, verbatim spans that never bisect a link/code span) and the opt-in `split_sentences_atomic` `SentenceSplitter`.

### `flowmark.ast`
- `walk_elements` (honest generic read-only tree walk) and `extract_links(doc) -> list[Link]` (`Link(text, url, title)`), reflecting what Markdown actually treats as a link (reference links resolved, images excluded by default, autolinks/bare URLs covered by the one `AutoLink` case). No span on `Link` — marko gives no inline offsets; span recovery is the consumer's source-mapping job.

### Behavior change
- Semantic line wrapping (`line_wrap_by_sentence`, the default `semantic=True` reformat path) is now **atomic-aware**: a sentence is never broken inside a link/code span/URL. This fixes a real bug where `St.` inside link text bisected the link across lines. Evaluated equal-or-better across the corpus (1 doc changed, 0 regressions); semantic/cleaned/auto goldens regenerated.

## Design notes
- **Identity vs. spans** — flowmark owns inline identity + heuristics; consumers map to source spans.
- **Heuristic vs. parser** — the atomic regexes are wrapping heuristics, not a link enumerator; use `extract_links` to enumerate links.

## Test plan
- [x] `iter_atomic_spans` round-trip + offset tests
- [x] `extract_links` identity cases (inline/reference/collapsed/duplicate/nested/escaped, autolinks, bare URLs, code-span & code-block exclusion)
- [x] `split_sentences_with_spans` verbatim + never-bisect-atomic
- [x] Wrapping golden corpus byte-identical after the word-splitter refactor (Phase B)
- [x] Semantic/cleaned/auto goldens regenerated for the Phase C default change
- [x] 329 pytest + 127 golden tryscript + lint all green

https://claude.ai/code/session_01GDGkKmpjMHhoDQTJQTwBqS

### CODE REVIEW

**Comment by jlevy** [comment]
## Senior engineering review

Summary: the high-level boundary is right: flowmark should expose reusable Markdown/formatting primitives, while chopdiff should own source-span reconciliation, `TextDoc`, sections, rollups, and link-to-sentence/document navigation. I would not merge the current PR as-is, though, because the PR has moved from plan-only to implementation while the public contract and implementation are still out of sync.

### Design boundary

The best split I see is:

- **flowmark** owns general primitives: configured Markdown parsing, parser-backed inline identity (`extract_links`), wrapper-safe atomic tokenization, and optionally sentence splitting with exact spans over raw text.
- **chopdiff** owns richer document semantics: the original source string, exact `[start, end)` mapping, block/section trees, link/span rollups, duplicate disambiguation, and any mapping from AST identity back onto source spans.

So I agree with not adding `Link.span` to flowmark. The docs should be careful not to imply that span recovery is just “locating the link text,” because duplicate link text, reference links, escaped text, and nested inline markup make that a chopdiff-level source-mapping problem, not a flowmark guarantee (`src/flowmark/ast.py:9`).

### Blocking issues

1. **The PR contract promises APIs that are not implemented.** The spec acceptance criteria and PR description promise `iter_atomic_tokens` and `split_sentences_with_spans`, but `src/flowmark/atomic.py` currently only re-exports patterns (`docs/project/specs/active/plan-2026-05-26-public-inline-api.md:104`, `src/flowmark/atomic.py:45`). Either split this into an explicitly Phase A PR and update the spec/PR body, or implement the promised token/span APIs before merge.

2. **`MARKDOWN_INLINE_PATTERNS` omits the URL/autolink cases that motivated the downstream need.** The spec says the Markdown-inline subset includes “autolinks/URLs,” but the implementation is exactly `(INLINE_CODE_SPAN, MARKDOWN_LINK)` and the test locks that in (`docs/project/specs/active/plan-2026-05-26-public-inline-api.md:154`, `src/flowmark/atomic.py:40`, `tests/test_public_inline_api.py:25`). That means `<https://...>` and GFM bare URLs are not atomic under the advertised prose pattern set. This matters directly for chopdiff’s Layer 4 goal: sentence spans can still bisect or mis-handle URL punctuation. Add explicit autolink/bare-URL patterns and tests, or narrow the claim.

3. **The public `AtomicPattern` type exposes an internal representation before the custom-pattern story is settled.** The plan says consumers may pass custom `AtomicPattern`s, but the re-exported dataclass requires `open_delim`, `close_delim`, `open_re`, and `close_re`, which are tag-handling internals rather than the general tokenization contract (`src/flowmark/atomic.py:19`, `src/flowmark/linewrapping/atomic_patterns.py:14`). Before stabilizing this API, consider a smaller public pattern type/factory with defaults, and adapt the current internal tag metadata behind it.

4. **`iter_inline` is not actually inline-only.** It yields every descendant with list children, including block elements, blank lines, and raw text inside fenced code blocks (`src/flowmark/ast.py:39`). That is surprising for a public helper named `iter_inline`, and it can make downstream consumers accidentally treat code-block content as inline document content. Either rename it to a generic tree iterator, or constrain it to inline scopes and explicitly skip code block bodies.

5. **The docs and PR metadata are stale now that implementation commits were added.** The PR body still says “Planning spec only” and “Docs-only,” while the spec status says “Planning. Not started.” (`docs/project/specs/active/plan-2026-05-26-public-inline-api.md:271`). Update the PR body and phase checklist so reviewers know whether they are reviewing a plan, Phase A code, or the whole promised API.

### Suggestions

- Keep top-level `flowmark.__all__` conservative. The submodules are a cleaner boundary; top-level re-exports should be limited to names you expect most callers to use directly.
- Add tests for nested link text (`[**bold** and \`code\`](url)`), shortcut/collapsed references, escaped brackets, bare GFM URLs, angle autolinks, duplicate link text, links inside code spans, and the `iter_inline` code-block exclusion behavior once that contract is decided.
- The current `gfm_elements.Url` branch in `extract_links` is effectively covered by `inline.AutoLink` because `Url` subclasses it. That is harmless, but the branch order/docstring should reflect the actual marko model.

### CI / local validation

- GitHub checks are passing on the current PR head across Python 3.10-3.14.
- I also ran `uv run pytest tests/test_public_inline_api.py` locally against the PR worktree; it passed.
- I checked chopdiff PR #8 for context; its CI is also green. Its proposed Layer 4 is exactly where flowmark should stop at primitives and let chopdiff own source spans and rollups.

File: src/flowmark/linewrapping/sentence_split_regex.py [hunk 1]
```diff
@@ -84,3 +93,67 @@ def first_sentence(
...
+
+class SentenceSpan(NamedTuple):
+    """A sentence with its exact `[start, end)` offsets into the source text."""
+
+    text: str
+    start: int
+    end: int
+
+
+def split_sentences_with_spans(
```
Comment (jlevy) [inline src/flowmark/linewrapping/sentence_split_regex.py:110]: Docs for this should clarify what "spans" means in terms of Markdown. This docstring doesn't mention markdown.

**Review by jlevy — COMMENTED** [review]
## Senior engineering review (current head `b0b7801`)

### Findings

1. **[P1] `extract_links` returns the destination, not the display text, for email autolinks.**
   `src/flowmark/ast.py:93-95` uses `Link(element.dest, element.dest, None)` for every `inline.AutoLink`. That is fine for `<https://example.com>`, but wrong for `<user@example.com>`: marko renders the link text as `user@example.com` and the destination as `mailto:user@example.com`. The public `Link.text` contract says this is the rendered/display text, so consumers will get a misleading identity for email autolinks. Suggested fix: use `_inline_text(element)` for the first field in the autolink branch and add coverage like `Link("user@example.com", "mailto:user@example.com", None)`.

2. **[P2] Empty link titles are collapsed into `None`.**
   `src/flowmark/ast.py:88-92` uses `element.title or None` for links/images. That loses the distinction between no title (`None`) and an explicitly empty Markdown title (`[x](url "")`, where marko sets `title == ""`). Since `Link.title` is already `str | None`, preserve the parser value with `element.title` directly. Add a small test for `[x](url "")` so the public API keeps exact link identity.

3. **[P2] `iter_atomic_spans(..., patterns=())` emits zero-length atomic spans.**
   `src/flowmark/linewrapping/atomic_patterns.py:240-263` compiles the empty pattern tuple as an empty regex, so `iter_atomic_spans("abc", patterns=())` yields zero-length `is_atomic=True` spans interleaved with the text. The docs say consumers may pass custom pattern tuples; an empty tuple is a natural “no atomics” case. Suggested fix: special-case `if not patterns: yield AtomicSpan(text, 0, len(text), False); return` (with the empty-string variant yielding nothing or one non-atomic empty span, whichever contract you prefer). It may also be worth rejecting custom regexes that can match the empty string.

4. **[P3] The completed spec still says the AST helpers are built on `transform_tree`, but the implementation uses a new generic walk.**
   `docs/project/specs/active/plan-2026-05-26-public-inline-api.md:95-97` and `:252-253` say `walk_elements` / `extract_links` use the existing `transform_tree`. `src/flowmark/ast.py:42-55` instead walks any element with list `children`, which is probably the better public behavior, but the spec should reflect the implementation so future maintainers do not “fix” it back to the narrower transformer path.

### Design assessment

The current design split is now in good shape: `flowmark.atomic` exposes heuristic unbreakable-span primitives, while `flowmark.ast` exposes parser-backed link identity and stays honest about not owning source offsets for marko inline nodes. I would keep that boundary. In particular, I would not add `Link.span` to flowmark; source reconciliation belongs in the downstream document model.

The Phase B refactor is also directionally right. Replacing placeholder substitution with `iter_atomic_words` gives the wrapper and public span API one source of truth. The Phase C default semantic-wrapping change looks justified by the golden diff: the changed Buckley paragraph is an improvement, not just churn.

### Suggested improvements

- Add tests for the two `extract_links` identity edge cases above: email autolinks and explicitly empty titles.
- Add one public-API test for `iter_atomic_spans(..., patterns=())` so the selectable-pattern contract is well-defined.
- Consider adding a short README or docs example before release showing the intended split: use `extract_links` for link identity, use `iter_atomic_spans` / `split_sentences_with_spans` for source-text spans, and do not enumerate links with the regex patterns.
- The existing inline PR thread about Markdown span terminology appears addressed by the newer docstrings, but it is still unresolved in the GitHub UI.

### Validation

- GitHub checks are passing on PR #47 for Python 3.10, 3.11, 3.12, 3.13, and 3.14.
- Local validation on the PR checkout passed:
  - `uv run python devtools/lint.py`
  - `uv run pytest` (`329 passed`)
  - `npx tryscript@latest run tests/tryscript/*.tryscript.md` (`127 passed`)

I would fix findings 1-3 before merge because this PR is creating a new public API surface, and these are cheap to correct now before downstream users depend on the edge-case behavior.

File: src/flowmark/linewrapping/sentence_split_regex.py (hunk 1 above)
Comment (jlevy) [reply src/flowmark/linewrapping/sentence_split_regex.py:110]: Addressed in b0b7801: `SentenceSpan` and `split_sentences_with_spans` now define a span as the sentence text plus its `[start, end)` offsets into the original Markdown source (`text == source[start:end]`, verbatim), and spell out which Markdown inline constructs (links, code spans, autolinks, bare URLs) are kept atomic. A Terminology block in the `flowmark.atomic` module docstring defines span/range/atomic/word/offsets so the API doesn't assume project context.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jlevy** [comment]
Reviewed this from the chopdiff side (the named downstream consumer) — it lands well for what chopdiff's block-aware `TextDoc` needs:

- `flowmark.atomic.split_sentences_with_spans` → verbatim, atomic-aware `SentenceSpan`s give chopdiff exact, link-safe sentence spans directly (this was the piece we'd otherwise have had to build).
- `flowmark.atomic` patterns + `iter_atomic_spans` cover exact link/code spans.
- `flowmark.ast.extract_links` + `walk_elements` cover correct link identity.

The identity-vs-spans boundary (links carry no span; consumer maps to its own source) is the right call — agreed, and chopdiff's plan owns that reconciliation.

One small, non-blocking ergonomics suggestion from doing that reconciliation: `AtomicSpan` carries `is_atomic` but not *which* pattern matched, so to attach spans to the links from `extract_links` a consumer has to re-match the link patterns to distinguish a link span from a code span. An optional matched-pattern `name` on `AtomicSpan` (or a `patterns=`-tagged variant of `iter_atomic_spans`) would make link-span reconciliation cheaper without crossing the identity/span boundary. Not needed for this PR.
